View change message handling

sergefdrv opened the pull request:
<!-- NOTE: Please check the contribution guideline before submitting -->

<!-- describe the purpose of the pull request, as well as its benefits
     and possible concerns related to the proposed changes -->
This pull request implements handling of the ViewChange message type.

<!-- and reference any issues resolved by the pull request, e.g. #123 -->
Resolves: #178

Commit:
core: Implement ViewChange message validation

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core: Implement ViewChange message processing

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

nhoriguchi commented on core/message-handling.go:
```diff
@@ -177,7 +177,9 @@ func defaultMessageHandlers(id uint32, log messagelog.MessageLog, unicastLogs ma
 	validateReqViewChange := makeReqViewChangeValidator(verifyMessageSignature)
-	validateCertifiedMessage := makeCertifiedMessageValidator(validatePrepare, validateCommit, verifyUI)
+	validateVCCert := makeViewChangeCertValidator(commitCertSize, n)
+	validateViewChange := makeViewChangeValidator(validateMessageLog, validateVCCert)
```
`validateMessageLog()` seems to be statically defined in `core/view-change.go`, so we don't have to explicitly pass the function via argument?

sergefdrv replied:
This dependency injection is useful for unit-testing

nhoriguchi replied:
OK, thanks.

nhoriguchi commented on core/view-change.go:
```diff
@@ -0,0 +1,130 @@
+
+func makeViewChangeCertValidator(commitCertSize, n uint32) viewChangeCertValidator {
+	return func(newView uint64, cert messages.ViewChangeCert) error {
+		if len(cert) <= int(n-commitCertSize) {
```
I guess that theoretical minimum requirement is `commitCertSize` and you intentionally use the stronger requirement of `n - commitCertSize + 1` based on discussion #215. If so,  is it worth adding comment (or leaving pointer to the #215 is also fine) about the choice?

sergefdrv replied:
Any view-change certificate must intersect with any commit certificate, otherwise delayed/slow replicas may keep triggering view change, thus the threshold of `n-commitCertSize+1`.

nhoriguchi replied:
Yes, so if this insight is from your own (not from the original paper), l thought that leaving some comment about it around the code is helpful for future developers.

sergefdrv replied:
Indeed, I will add some comment

nhoriguchi commented on core/view-change.go:
```diff
@@ -128,3 +153,71 @@ func validateMessageLog(replicaID uint32, view uint64, log messages.MessageLog,
+
+		if vc.ReplicaID() == id {
+			collected[0] = vc
+		} else if len(collected) <= int(n-commitCertSize) {
```
This kind of check is called multiple times, so how about factoring out to a separate function with a descriptive name?

sergefdrv replied:
Maybe just define `threshold := n-commitCertSize`?

sergefdrv replied:
I think the condition here should be `len(collected) < int(n-commitCertSize)` resp. `len(collected) < int(threshold)`, need to check the unit-test...

sergefdrv replied:
Ah, no, `commitCertSize == f+1`, so the condition is correct. Then we would define `threshold := n-commitCertSize+1`.

nhoriguchi replied:
Thank you for double checking, `threshold := n-commitCertSize` looks fine to me.

sergefdrv replied:
Perhaps `viewChangeCertSize` would be a better name. What do you think about passing it to `makeViewChangeCertValidator`, `makeViewChangeCollector`, and `makeReqViewChangeCollector` as a parameter?

nhoriguchi replied:
Sounds nice 👍

sergefdrv commented on core/view-change.go:
```diff
@@ -128,3 +153,71 @@ func validateMessageLog(replicaID uint32, view uint64, log messages.MessageLog,
+func makeViewChangeCollector(id, commitCertSize, n uint32) viewChangeCollector {
+	var (
+		view      uint64
+		collected = make(messages.NewViewCert, 1, n-commitCertSize+1)
```
Oh, the capacity should be `n-commitCertSize`

sergefdrv replied:
The capacity is correct, see [here](https://github.com/hyperledger-labs/minbft/pull/230/files/45aa5e4aa2cf5a145e64a9af6c67b366cab016c1..db5ca76d3b2750be3a942fe9bd1285d21b227898#r789942736).

Commit:
core: Define view-change certificate size explicitly

Define a variable to hold the view-change certificate size next to
commitCertSize definition and pass it to makeReqViewChangeCollector
explicitly.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

sergefdrv commented:
Summary of [changes](https://github.com/hyperledger-labs/minbft/compare/db5ca76d3b2750be3a942fe9bd1285d21b227898..4103c85aa7e8cbeeb14be04142044ef177e47812): added explicit definition for view-change certificate size.

nhoriguchi reviewed (approved):
Thank you very much.
